File: Python/InformasiFilm.py
# =====================================================================
# FILE   : InformasiFilm.py
# PERAN  : BASE CLASS 1 (salah satu "orang tua" dari class Penayangan)
# ISI    : Menyimpan data yang berhubungan dengan FILM saja:
#          judul, genre, dan durasi.
# ALUR   : Class ini diimpor dan diwarisi oleh Penayangan.py
#          (class Penayangan(InformasiFilm, InformasiJadwal): ...)
#
# CATATAN PYTHON vs C++/JAVA:
#   - Python TIDAK butuh include guard / import guard: Python otomatis
#     hanya menjalankan isi sebuah file (module) satu kali walau
#     di-import dari banyak tempat.
#   - Python (berbeda dari Java) mengizinkan multiple inheritance secara
#     langsung, jadi struktur pewarisannya bisa persis sama seperti versi
#     C++ (Penayangan mewarisi InformasiFilm DAN InformasiJadwal sekaligus).
# =====================================================================

class InformasiFilm:
    """Merepresentasikan data film yang sedang tayang di bioskop."""

    # ---- Constructor ----
    # Python hanya punya SATU constructor per class, bernama __init__.
    # Untuk meniru "constructor default" (C++/Java) DAN "constructor
    # berparameter" sekaligus, ketiga parameter diberi nilai default
    # ("" dan 0). Jadi InformasiFilm() dan InformasiFilm("A","B",1)
    # sama-sama valid, seperti overloading constructor di bahasa lain.
    def __init__(self, judul_film="", genre="", durasi_menit=0):
        # Semua atribut di Python otomatis bisa diakses layaknya
        # 'protected' di C++ (konvensi: tidak ada penanda private/
        # protected/public seperti C++/Java; garis bawah tunggal di depan
        # nama, mis. "_x", HANYA konvensi "jangan diakses dari luar",
        # bukan larangan sesungguhnya). Di sini nama dibiarkan polos
        # karena class turunan memang perlu mengaksesnya langsung.
        self.judul_film = judul_film
        self.genre = genre
        self.durasi_menit = durasi_menit

    # ---- Setter & getter ----
    # Sebenarnya di Python atribut bisa diakses langsung (obj.judul_film),
    # tetapi getter/setter tetap dibuat di sini agar strukturnya sama
    # persis dengan versi C++ dan Java, dan agar Main.py bisa memanggil
    # obj.getJudulFilm() dst. seperti pada kedua versi sebelumnya.
    def setJudulFilm(self, judul_film):
        self.judul_film = judul_film

    def getJudulFilm(self):
        return self.judul_film

    def setGenre(self, genre):
        self.genre = genre

    def getGenre(self):
        return self.genre

    def setDurasiMenit(self, durasi_menit):
        self.durasi_menit = durasi_menit

    def getDurasiMenit(self):
        return self.durasi_menit

File: PHP/classes/InformasiFilm.php
<?php
// =====================================================================
// FILE   : classes/InformasiFilm.php
// PERAN  : BASE CLASS 1 (parent dari class Penayangan)
// ISI    : Menyimpan data yang berhubungan dengan FILM saja:
//          judul, genre, durasi, DAN (baru) foto/poster film.
// ALUR   : Class ini di-extends oleh Penayangan.php.
//
// ATRIBUT BARU: fotoFilm
//   Menyimpan NAMA FILE poster film yang sudah diupload user (bukan
//   data gambar itu sendiri). File asli disimpan fisik di folder
//   uploads/, sedangkan yang disimpan di objek/session hanyalah nama
//   filenya, contoh: "foto_65f2a1.jpg". Saat ditampilkan, tag <img>
//   di index.php akan membaca file itu dari folder uploads/.
//   Ditempatkan di InformasiFilm (bukan di Penayangan) karena secara
//   konsep, poster adalah atribut milik FILM-nya, bukan milik jadwal
//   tayangnya (satu film yang sama tayang di banyak jadwal tetap
//   pakai poster yang sama).
// =====================================================================

class InformasiFilm {

    // 'protected' = tidak bisa diakses dari luar class, tetapi BISA
    // diakses langsung oleh class turunan (Penayangan dst.), sama
    // seperti versi C++ dan Java.
    protected $judulFilm;
    protected $genre;
    protected $durasiMenit;
    protected $fotoFilm;   // nama file poster, contoh: "foto_65f2a1.jpg" atau "" jika tidak ada

    // ---- Constructor ----
    // PHP hanya punya SATU constructor per class (__construct), sama
    // seperti Python. Nilai default (= "", = 0) meniru constructor
    // default C++/Java sekaligus constructor berparameter dalam satu method.
    public function __construct($judulFilm = "", $genre = "", $durasiMenit = 0, $fotoFilm = "") {
        $this->judulFilm = $judulFilm;
        $this->genre = $genre;
        $this->durasiMenit = $durasiMenit;
        $this->fotoFilm = $fotoFilm;
    }

    // ---- Setter & getter ----
    public function setJudulFilm($judulFilm) { $this->judulFilm = $judulFilm; }
    public function getJudulFilm() { return $this->judulFilm; }

    public function setGenre($genre) { $this->genre = $genre; }
    public function getGenre() { return $this->genre; }

    public function setDurasiMenit($durasiMenit) { $this->durasiMenit = $durasiMenit; }
    public function getDurasiMenit() { return $this->durasiMenit; }

    // Setter/getter khusus foto. getFotoFilm() dipakai di index.php untuk
    // menentukan path gambar yang ditampilkan (lihat komentar di index.php).
    public function setFotoFilm($fotoFilm) { $this->fotoFilm = $fotoFilm; }
    public function getFotoFilm() { return $this->fotoFilm; }

    // Method bantu: true jika film ini punya foto yang tersimpan.
    // Dipakai index.php untuk memilih antara <img> asli atau placeholder.
    public function adaFoto() {
        return $this->fotoFilm !== "" && $this->fotoFilm !== null;
    }
}
